File: GoTracer/src/Types/Scene.go
package types

import (
	"image/color"
	"math"
)

type Scene struct {
	Objects []SceneObject
	Lights  []Light
	Camera  Camera
}

func (scene Scene) intersection(ray Ray) *Intersection {
	var intersection *Intersection = nil
	for index := range scene.Objects {
		object := scene.Objects[index]
		isect := object.Intersects(ray)
		if isect != nil {
			if intersection == nil {
				intersection = isect
			} else if isect.Distance < intersection.Distance {
				intersection = isect
			}
		}
	}
	return intersection
}

func (scene Scene) testRay(ray Ray) float64 {
	intersection := scene.intersection(ray)
	if intersection != nil {
		return intersection.Distance
	} else {
		return 0
	}
}

func (scene Scene) traceRay(ray Ray, depth int) Color {
	black := Color{R: 0.0, G: 0.0, B: 0.0}
	isect := scene.intersection(ray)
	if isect == nil {
		return black
	}
	return scene.shade(*isect, depth)
}

func (scene Scene) shade(intersection Intersection, depth int) Color {
	pos := intersection.Ray.Direction.Scale(intersection.Distance).Add(intersection.Ray.Position)
	normal := intersection.Object.Normal(pos)
	dot := normal.Dot(intersection.Ray.Direction)
	reflectDir := intersection.Ray.Direction.Subtract(normal.Scale(2 * dot))
	natural := scene.getNaturalColor(intersection.Object, pos, normal, reflectDir)
	if depth < 5 {
		reflectPos := reflectDir.Scale(0.001).Add(pos)
		reflectColor := scene.getReflectionColor(intersection.Object, reflectPos, reflectDir, depth)
		return natural.Add(reflectColor)
	}
	return Color{R: 0.5, G: 0.5, B: 0.5}.Add(natural)
}

func (scene Scene) getNaturalColor(object SceneObject, position Vector, normal Vector, rayDirection Vector) Color {
	var result Color
	black := Color{R: 0.0, G: 0.0, B: 0.0}
	for index := range scene.Lights {
		light := scene.Lights[index]
		lightDistance := light.Position.Subtract(position)
		livec := lightDistance.Normalize()
		ray := Ray{Position: position, Direction: livec}
		isect := scene.testRay(ray)
		isInShadow := (isect > lightDistance.Length() || isect == 0.0)
		if isInShadow {
			illum := livec.Dot(normal)

			lcolor := black
			if illum > 0.0 {
				lcolor = light.Color.Scale(illum)
			}
			normalizedRayDirection := rayDirection.Normalize()
			specular := livec.Dot(normalizedRayDirection)
			scolor := black
			if specular > 0.0 {
				scolor = light.Color.Scale(math.Pow(specular, object.GetSurface().Roughness()))
			}

			diffuseResult := object.GetSurface().Diffuse(position).Multiply(lcolor)
			specularResult := object.GetSurface().Specular(position).Multiply(scolor)

			result = result.Add(diffuseResult).Add(specularResult)
		}
	}
	return result
}

func (scene Scene) getReflectionColor(object SceneObject, pos Vector, rayDirection Vector, depth int) Color {
	ray := Ray{Position: pos, Direction: rayDirection}
	return scene.traceRay(ray, depth+1).Scale(object.GetSurface().Reflect(pos))
}

func recenter_x(x int, halfWidth float64, invWidth float64) float64 {
	dx := float64(x)
	return (dx - halfWidth) * invWidth
}

func recenter_y(y int, halfHeight float64, invHeight float64) float64 {
	dy := float64(y)
	return -(dy - halfHeight) * invHeight
}

func getPoint(x int, y int, halfWidth float64, halfHeight float64, invWidth float64, invHeight float64, camera Camera) Vector {
	right := camera.Right.Scale(recenter_x(x, halfWidth, invWidth))
	up := camera.Up.Scale(recenter_y(y, halfHeight, invHeight))
	return right.Add(up).Add(camera.Forward).Normalize()
}

func (scene Scene) Render(width int, height int) []color.Color {
	fWidth := float64(width)
	fHeight := float64(height)
	inverseAspectRatio := fHeight / fWidth
	scaleWidth := fWidth * inverseAspectRatio
	halfWidth := scaleWidth / 2.0
	halfHeight := fHeight / 2.0
	invWidth := 1.0 / (scaleWidth * 2.0)
	invHeight := 1.0 / (fHeight * 2.0)

	result := make([]color.Color, width*height)
	basey := 0
	for y := 0; y < height; y++ {
		basey = y * width
		for x := 0; x < width; x++ {
			rayDir := getPoint(x, y, halfWidth, halfHeight, invWidth, invHeight, scene.Camera)
			ray := Ray{Position: scene.Camera.Position, Direction: rayDir}
			result[basey+x] = scene.traceRay(ray, 0).ToColor()
		}
	}
	return result
}

File: RustTracer/Raytracer/scene.rs
use raytracer::camera::Camera;
use raytracer::scene_object::SceneObject;
use raytracer::light::Light;
use raytracer::color::Color;
use raytracer::vector::Vector;
use raytracer::ray::Ray;
use raytracer::intersection::Intersection;
use std::num::pow;

pub struct Scene<'a> {
	pub camera : Camera,
	pub objects : Vec<&'a SceneObject<'a> +'a> ,
	pub lights : Vec<Light>
}

impl<'a> Scene<'a> {
	fn intersection(&'a self, ray : Ray) -> Option<(Intersection, &SceneObject<'a>)> {
		self.objects.iter().fold(None, |intersection: Option<(Intersection, _)>, object| {
			let isect = object.intersects(ray);
			match (isect, intersection) {
				(Some((i1,_)), Some((i2,_))) => if i1.distance < i2.distance { isect } else { intersection }, 
				(Some(_), None) => isect,
				(None, _) => intersection
			}})
	}

	fn test_ray(&'a self, ray:Ray) -> f64 {
		let intersection = self.intersection(ray);
		match intersection {
			Some((isect,_)) => isect.distance,
			None => 0.0
		}
	}

	fn trace_ray(&'a self, ray:Ray, depth:int) -> Color { 
		let black = Color{ r: 0.0, g: 0.0, b: 0.0 };
		let isect = self.intersection(ray);
		match isect {
			Some(isect) => self.shade(isect, depth),
			None => black
		}
	}

	fn shade(&'a self,intersection : (Intersection, &'a SceneObject<'a>), depth : int) -> Color { 
		let (intersection, object) = intersection;
		let pos = intersection.ray.direction.scale(intersection.distance).add(intersection.ray.position);
		let normal = object.normal(pos);
		let dot = normal.dot(intersection.ray.direction);	
		let reflect_dir = intersection.ray.direction.subtract(normal.scale(2.0 * dot));
		let natural = self.get_natural_color(object, pos, normal, reflect_dir);
		if depth < 5 { 
			let reflect_pos = reflect_dir.scale(0.001).add(pos);
			let reflect_color = self.get_reflection_color(object, reflect_pos, reflect_dir, depth);
			reflect_color.add(natural)
		} else { 
			Color { r: 0.5, g: 0.5, b: 0.5}.add(natural)
		}
	}

	fn get_natural_color(&'a self, object : &'a SceneObject<'a>, position : Vector, normal : Vector, ray_direction : Vector ) -> Color {
		let black = Color{r: 0.0, g: 0.0, b: 0.0};
		self.lights.iter().fold(black, |result, light|  {
			let light_distance = light.position.subtract(position);
			let livec = light_distance.normalize();
			let ray = Ray{position: position, direction: livec};
			let isect = self.test_ray(ray);
			let is_in_shadow = isect > light_distance.length() || isect == 0.0;
			if is_in_shadow {
				let illum = livec.dot(normal);
				let lcolor = 
					if illum > 0.0 {
						light.color.scale(illum)
					} else { black };
				let normalized_ray_direction = ray_direction.normalize();
				let specular = livec.dot(normalized_ray_direction);
				let scolor = 
					if specular > 0.0 {
						light.color.scale(pow(specular, object.get_surface().roughness()))
					} else { black };

				let diffuse_result = object.get_surface().diffuse(position).multiply(lcolor);
				let specular_result = object.get_surface().specular(position).multiply(scolor);

				result.add(diffuse_result).add(specular_result)
			} else {
				result
			}
		})
	}

	fn get_reflection_color(&'a self, object : &'a SceneObject<'a>, pos:Vector, ray_direction : Vector, depth :int ) -> Color {
		let ray = Ray { position : pos, direction : ray_direction };
		(self.trace_ray(ray, depth + 1)).scale(object.get_surface().reflect(pos))
	}

	fn recenter_x(&self, x :uint, half_width :f64, inv_width :f64) -> f64 {
		let dx = x as f64;
		(dx - half_width) * inv_width
	}

	fn recenter_y(&self, y : uint, half_height : f64, inv_height : f64) -> f64 {
		let dy = y as f64;
		(dy - half_height) * inv_height //note removed negation
	}

	fn get_point(&self, x : uint, y : uint, half_width : f64, half_height : f64, inv_width : f64, inv_height : f64, camera : Camera) -> Vector {
		let right = camera.right.scale(self.recenter_x(x, half_width, inv_width));
		let up = camera.up.scale(self.recenter_y(y, half_height, inv_height));
		right.add(up).add(camera.forward).normalize()
	}

	pub fn render(&'a self, width : uint, height : uint) -> Vec<int> {
		let f_width = width as f64;
		let f_height = height as f64;
		let inverse_aspect_ratio = f_height / f_width;
		let scale_width = f_width * inverse_aspect_ratio;
		let half_width = scale_width / 2.0;
		let half_height = f_height / 2.0;
	    let inv_width = 1.0 / (scale_width * 2.0);
		let inv_height = 1.0 / (f_height * 2.0);
		let array_size = width * height;
		let mut result : Vec<int> = Vec::with_capacity(array_size);
		for y in range(0,height) {
			for x in range(0,width) {
				let ray_dir = self.get_point(x, y, half_width, half_height, inv_width, inv_height, self.camera);
				let ray = Ray{position: self.camera.position, direction: ray_dir};
				let col = self.trace_ray(ray, 0).to_color();
				result.push(col);
			}
		}
		result
	}
}